src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+    // Store in watcherInfo for later use if we need to reset
+    // things after an 'end' caught here.
+    let options = watcherInfo.watchmanWatcher.createOptions();  
+    watcherInfo.options = options;
+    
+    // Dup the options object so we can add 'relative_root' and 'since'
+    // and leave the original options object alone. We'll do this again
+    // later if we need to resubscribe after 'end' and reconnect.
+    options = Object.assign({}, options);
+
+    if (this._relative_root) {
+      options.relative_root = watcherInfo.relativePath;
+    }
+
+    options.since = watcherInfo.since;
+
+    this._client.command(['subscribe', watcherInfo.root, subscription, options],
+                         (error, resp) => {
+                           if (error) {
+                             reject(error);
+                           } else {
+                             resolve(resp);
+                           }
+                         });
+  });
+  
+  return promise;
+}
+
+/**
+ * Handle the 'subscription' (file change) event, by calling the 
+ * handler on the relevant WatchmanWatcher.
+ */
+WatchmanClient.prototype._onSubscription = function(resp) {
+  let watcherInfo = this._getWatcherInfo(resp.subscription);
+  if (watcherInfo) {
+    watcherInfo.watchmanWatcher.handleChangeEvent(resp);
+  } else {
+    // Note it in the log, but otherwise ignore it
+    console.error("WatchmanClient error - received 'subscription' event " +
+                  "for non-existent subscription '" +
+                  resp.subscription + "'");
+  }
+};
+
+/**
+ * Handle the 'error' event by forwarding to the
+ * handler on the relevant WatchmanWatcher.
+ */
+WatchmanClient.prototype._onError = function(error) {
+  Object.values(this._watcherMap).forEach((watcherInfo) =>
+                                          watcherInfo.watchmanWatcher.handleErrorEvent(error));
+};
+
+/**
+ * Handle the 'end' event by creating a new watchman.Client and
+ * attempting to resubscribe all the existing subscriptions, but
+ * without notifying the WatchmanWatchers about it. They should
+ * not be aware the connection was lost and recreated.
+ * If something goes wrong during any part of the reconnect/setup,
+ * call the error handler on each existing WatchmanWatcher.
+ */
+WatchmanClient.prototype._onEnd = function() {
+  console.warn('[sane.WatchmanClient] Warning: Lost connection to watchman, reconnecting..');
+
+  // Hold the old watcher map so we use it to recreate all subscriptions.
+  // Hold the old watchmanBinaryPath because it should not have changed.
+  let oldWatcherInfos = Object.values(this._watcherMap);
+  let watchmanBinaryPath = this._watchmanBinaryPath;
+
+  clearLocalVars();
+  
+  this.getClient(watchmanBinaryPath)
+    .then(
+      (client) => {
+        let promises = oldWatcherInfos.map((watcherInfo) =>
+                                           subscribe(watcherInfo.watchmanWatcher,
```

Comment:
Where is this `subscribe` defined, I believe this may be a refactoring error.

Reply:
Good catch! I'd done a manual test to close the connection before I did the refactoring, and hadn't yet repeated it for the new version. Fixed.